### ISSUE TITLE
need to figure out why cpm fails

### ISSUE THREAD

**Issue by skaji** [skaji/cpm#66]
cc @eserte

As reported in #51, cpm sometimes fails to install some modules, especially with old perls.

I confirmed that cpm failed to install $module with perl 5.10:

Cpanel::JSON::XS (via `cpm install Plack`)
```
2016-12-23T18:27:00,25429,Cpanel-JSON-XS-3.0225| Unpacking Cpanel-JSON-XS-3.0225.tar.gz
2016-12-23T18:27:00,49090,Cpanel-JSON-XS-3.0225| Checking if your kit is complete...
2016-12-23T18:27:00,49090,Cpanel-JSON-XS-3.0225| Looks good
2016-12-23T18:27:00,49090,Cpanel-JSON-XS-3.0225| Generating a Unix-style Makefile
2016-12-23T18:27:00,49090,Cpanel-JSON-XS-3.0225| Writing Makefile for Cpanel::JSON::XS
2016-12-23T18:27:00,49090,Cpanel-JSON-XS-3.0225| Writing MYMETA.yml and MYMETA.json
2016-12-23T18:27:01,49884,Cpanel-JSON-XS-3.0225| cp XS/Boolean.pm blib/lib/Cpanel/JSON/XS/Boolean.pm
2016-12-23T18:27:01,49884,Cpanel-JSON-XS-3.0225| cp XS.pm blib/lib/Cpanel/JSON/XS.pm
2016-12-23T18:27:01,49884,Cpanel-JSON-XS-3.0225| Running Mkbootstrap for XS ()
2016-12-23T18:27:01,49884,Cpanel-JSON-XS-3.0225| chmod 644 "XS.bs"
2016-12-23T18:27:01,49884,Cpanel-JSON-XS-3.0225| "/Users/skaji/env/plenv/versions/5.10.1/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- XS.bs blib/arch/auto/Cpanel/JSON/XS/XS.bs 644
2016-12-23T18:27:01,49884,Cpanel-JSON-XS-3.0225| "/Users/skaji/env/plenv/versions/5.10.1/bin/perl" "/Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/ExtUtils/xsubpp"  -typemap '/Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/ExtUtils/typemap' -typemap '/Users/skaji/.perl-cpm/work/Cpanel-JSON-XS-3.0225/typemap'  XS.xs > XS.xsc
2016-12-23T18:27:01,49884,Cpanel-JSON-XS-3.0225| Undefined subroutine &ExtUtils::ParseXS::errors called at /Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/ExtUtils/xsubpp line 41.
2016-12-23T18:27:01,49884,Cpanel-JSON-XS-3.0225| make: *** [XS.c] Error 255
```
https://gist.github.com/skaji/0b4c5d5acefd17f95b9ab08fedd462ab

 HTTP::Entity::Parser (via `cpm install Plack`)
```
2016-12-23T18:32:40,74375,HTTP::Entity::Parser| Resolved HTTP::Entity::Parser (0.17) -> http://www.cpan.org/authors/id/K/KA/KAZEBURO/HTTP-Entity-Parser-0.18.tar.gz from MetaDB
2016-12-23T18:32:40,74375,HTTP-Entity-Parser-0.18| Unpacking HTTP-Entity-Parser-0.18.tar.gz
2016-12-23T18:32:40,81529,HTTP-Entity-Parser-0.18| Can't locate CPAN/Meta/Prereqs.pm in @INC (@INC contains: /Users/skaji/work/2016-12-18/local/lib/perl5/5.10.1/darwin-2level /Users/skaji/work/2016-12-18/local/lib/perl5/5.10.1 /Users/skaji/work/2016-12-18/local/lib/perl5/darwin-2level /Users/skaji/work/2016-12-18/local/lib/perl5 /Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/darwin-2level /Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1 /Users/skaji/env/plenv/versions/5.10.1/lib/perl5/site_perl/5.10.1/darwin-2level /Users/skaji/env/plenv/versions/5.10.1/lib/perl5/site_perl/5.10.1 .) at /Users/skaji/work/2016-12-18/local/lib/perl5/CPAN/Meta/Feature.pm line 8.
2016-12-23T18:32:40,81529,HTTP-Entity-Parser-0.18| BEGIN failed--compilation aborted at /Users/skaji/work/2016-12-18/local/lib/perl5/CPAN/Meta/Feature.pm line 8.
2016-12-23T18:32:40,81529,HTTP-Entity-Parser-0.18| Compilation failed in require at /Users/skaji/work/2016-12-18/local/lib/perl5/CPAN/Meta.pm line 57.
2016-12-23T18:32:40,81529,HTTP-Entity-Parser-0.18| BEGIN failed--compilation aborted at /Users/skaji/work/2016-12-18/local/lib/perl5/CPAN/Meta.pm line 57.
2016-12-23T18:32:40,81529,HTTP-Entity-Parser-0.18| Compilation failed in require at /Users/skaji/work/2016-12-18/local/lib/perl5/Module/Build/Tiny.pm line 8.
2016-12-23T18:32:40,81529,HTTP-Entity-Parser-0.18| BEGIN failed--compilation aborted at /Users/skaji/work/2016-12-18/local/lib/perl5/Module/Build/Tiny.pm line 8.
2016-12-23T18:32:40,81529,HTTP-Entity-Parser-0.18| Compilation failed in require at Build.PL line 9.
2016-12-23T18:32:40,81529,HTTP-Entity-Parser-0.18| BEGIN failed--compilation aborted at Build.PL line 9.
```

I'm not sure, but...
* local::lib issue ?
* the order of `@INC` ?

**Comment by skaji** [comment]
Command: /Users/skaji/env/plenv/versions/5.10.1/bin/perl -I/Users/skaji/work/2016-12-18/local510/lib/perl5 -Ilib script/cpm install Plack
Current directory: `/Users/skaji/src/github.com/skaji/cpm`

case OK -> using /Users/skaji/src/github.com/skaji/cpm/local/lib/perl5/ExtUtils/xsubpp
```
> grep Cpanel ~/.perl-cpm/build.log.1482490471
2016-12-23T19:54:52,87111,JSON-MaybeXS-1.003008| Warning: prerequisite Cpanel::JSON::XS 2.3310 not found.
2016-12-23T19:54:52,65312,JSON-MaybeXS-1.003008| Found dependencies: Cpanel::JSON::XS (2.3310)
2016-12-23T19:54:53,65321,Cpanel::JSON::XS| Resolved Cpanel::JSON::XS (2.3310) -> http://www.cpan.org/authors/id/R/RU/RURBAN/Cpanel-JSON-XS-3.0225.tar.gz from 02Packages
2016-12-23T19:54:53,65321,Cpanel-JSON-XS-3.0225| Unpacking Cpanel-JSON-XS-3.0225.tar.gz
2016-12-23T19:54:53,89140,Cpanel-JSON-XS-3.0225| Checking if your kit is complete...
2016-12-23T19:54:53,89140,Cpanel-JSON-XS-3.0225| Looks good
2016-12-23T19:54:53,89140,Cpanel-JSON-XS-3.0225| Generating a Unix-style Makefile
2016-12-23T19:54:53,89140,Cpanel-JSON-XS-3.0225| Writing Makefile for Cpanel::JSON::XS
2016-12-23T19:54:53,89140,Cpanel-JSON-XS-3.0225| Writing MYMETA.yml and MYMETA.json
2016-12-23T19:54:54,89700,Cpanel-JSON-XS-3.0225| cp XS/Boolean.pm blib/lib/Cpanel/JSON/XS/Boolean.pm
2016-12-23T19:54:54,89700,Cpanel-JSON-XS-3.0225| cp XS.pm blib/lib/Cpanel/JSON/XS.pm
2016-12-23T19:54:54,89700,Cpanel-JSON-XS-3.0225| Running Mkbootstrap for XS ()
2016-12-23T19:54:54,89700,Cpanel-JSON-XS-3.0225| chmod 644 "XS.bs"
2016-12-23T19:54:54,89700,Cpanel-JSON-XS-3.0225| "/Users/skaji/env/plenv/versions/5.10.1/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- XS.bs blib/arch/auto/Cpanel/JSON/XS/XS.bs 644
2016-12-23T19:54:54,89700,Cpanel-JSON-XS-3.0225| "/Users/skaji/env/plenv/versions/5.10.1/bin/perl" "/Users/skaji/src/github.com/skaji/cpm/local/lib/perl5/ExtUtils/xsubpp"  -typemap '/Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/ExtUtils/typemap' -typemap '/Users/skaji/.perl-cpm/work/Cpanel-JSON-XS-3.0225/typemap'  XS.xs > XS.xsc
2016-12-23T19:54:54,89700,Cpanel-JSON-XS-3.0225| mv XS.xsc XS.c
2016-12-23T19:54:54,89700,Cpanel-JSON-XS-3.0225| cc -c   -fno-common -DPERL_DARWIN -mmacosx-version-min=10.12 -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -O3   -DVERSION=\"3.0225\" -DXS_VERSION=\"3.0225\"  "-I/Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/darwin-2level/CORE"   XS.c
2016-12-23T19:54:55,89700,Cpanel-JSON-XS-3.0225| rm -f blib/arch/auto/Cpanel/JSON/XS/XS.bundle
2016-12-23T19:54:55,89700,Cpanel-JSON-XS-3.0225| cc  -mmacosx-version-min=10.12 -bundle -undefined dynamic_lookup -L/usr/local/lib -fstack-protector XS.o  -o blib/arch/auto/Cpanel/JSON/XS/XS.bundle  \
2016-12-23T19:54:55,89700,Cpanel-JSON-XS-3.0225| 	      \
2016-12-23T19:54:55,89700,Cpanel-JSON-XS-3.0225|
2016-12-23T19:54:55,89700,Cpanel-JSON-XS-3.0225| chmod 755 blib/arch/auto/Cpanel/JSON/XS/XS.bundle
2016-12-23T19:54:55,89700,Cpanel-JSON-XS-3.0225| cp bin/cpanel_json_xs blib/script/cpanel_json_xs
2016-12-23T19:54:55,89700,Cpanel-JSON-XS-3.0225| "/Users/skaji/env/plenv/versions/5.10.1/bin/perl" -MExtUtils::MY -e 'MY->fixin(shift)' -- blib/script/cpanel_json_xs
2016-12-23T19:54:55,65321,Cpanel-JSON-XS-3.0225| "/Users/skaji/env/plenv/versions/5.10.1/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- XS.bs blib/arch/auto/Cpanel/JSON/XS/XS.bs 644
2016-12-23T19:54:56,65321,Cpanel-JSON-XS-3.0225| Files found in blib/arch: installing files in blib/lib into architecture dependent library tree
2016-12-23T19:54:56,65321,Cpanel-JSON-XS-3.0225| Installing /Users/skaji/src/github.com/skaji/cpm/local/lib/perl5/darwin-2level/auto/Cpanel/JSON/XS/XS.bundle
2016-12-23T19:54:56,65321,Cpanel-JSON-XS-3.0225| Installing /Users/skaji/src/github.com/skaji/cpm/local/lib/perl5/darwin-2level/Cpanel/JSON/XS.pm
2016-12-23T19:54:56,65321,Cpanel-JSON-XS-3.0225| Installing /Users/skaji/src/github.com/skaji/cpm/local/lib/perl5/darwin-2level/Cpanel/JSON/XS/Boolean.pm
2016-12-23T19:54:56,65321,Cpanel-JSON-XS-3.0225| Installing /Users/skaji/src/github.com/skaji/cpm/local/bin/cpanel_json_xs
2016-12-23T19:54:56,65321,Cpanel-JSON-XS-3.0225| Appending installation info to /Users/skaji/src/github.com/skaji/cpm/local/lib/perl5/darwin-2level/perllocal.pod
2016-12-23T19:54:56,65321,Cpanel-JSON-XS-3.0225| Installing /Users/skaji/src/github.com/skaji/cpm/local/lib/perl5/darwin-2level/.meta/Cpanel-JSON-XS-3.0225/install.json
2016-12-23T19:54:56,65321,Cpanel-JSON-XS-3.0225| Installing /Users/skaji/src/github.com/skaji/cpm/local/lib/perl5/darwin-2level/.meta/Cpanel-JSON-XS-3.0225/MYMETA.json
```

case NG -> using /Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/ExtUtils/xsubp
```
> grep Cpanel ~/.perl-cpm/build.log.1482490504
2016-12-23T19:55:17,20480,JSON-MaybeXS-1.003008| Warning: prerequisite Cpanel::JSON::XS 2.3310 not found.
2016-12-23T19:55:17,97676,JSON-MaybeXS-1.003008| Found dependencies: Cpanel::JSON::XS (2.3310)
2016-12-23T19:55:17,97682,Cpanel::JSON::XS| Resolved Cpanel::JSON::XS (2.3310) -> http://www.cpan.org/authors/id/R/RU/RURBAN/Cpanel-JSON-XS-3.0225.tar.gz from 02Packages
2016-12-23T19:55:17,97682,Cpanel-JSON-XS-3.0225| Unpacking Cpanel-JSON-XS-3.0225.tar.gz
2016-12-23T19:55:17,21667,Cpanel-JSON-XS-3.0225| Checking if your kit is complete...
2016-12-23T19:55:17,21667,Cpanel-JSON-XS-3.0225| Looks good
2016-12-23T19:55:17,21667,Cpanel-JSON-XS-3.0225| Generating a Unix-style Makefile
2016-12-23T19:55:17,21667,Cpanel-JSON-XS-3.0225| Writing Makefile for Cpanel::JSON::XS
2016-12-23T19:55:17,21667,Cpanel-JSON-XS-3.0225| Writing MYMETA.yml and MYMETA.json
2016-12-23T19:55:18,22431,Cpanel-JSON-XS-3.0225| cp XS/Boolean.pm blib/lib/Cpanel/JSON/XS/Boolean.pm
2016-12-23T19:55:18,22431,Cpanel-JSON-XS-3.0225| cp XS.pm blib/lib/Cpanel/JSON/XS.pm
2016-12-23T19:55:18,22431,Cpanel-JSON-XS-3.0225| Running Mkbootstrap for XS ()
2016-12-23T19:55:18,22431,Cpanel-JSON-XS-3.0225| chmod 644 "XS.bs"
2016-12-23T19:55:18,22431,Cpanel-JSON-XS-3.0225| "/Users/skaji/env/plenv/versions/5.10.1/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- XS.bs blib/arch/auto/Cpanel/JSON/XS/XS.bs 644
2016-12-23T19:55:18,22431,Cpanel-JSON-XS-3.0225| "/Users/skaji/env/plenv/versions/5.10.1/bin/perl" "/Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/ExtUtils/xsubpp"  -typemap '/Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/ExtUtils/typemap' -typemap '/Users/skaji/.perl-cpm/work/Cpanel-JSON-XS-3.0225/typemap'  XS.xs > XS.xsc
2016-12-23T19:55:18,22431,Cpanel-JSON-XS-3.0225| Undefined subroutine &ExtUtils::ParseXS::errors called at /Users/skaji/env/plenv/versions/5.10.1/lib/perl5/5.10.1/ExtUtils/xsubpp line 41.
2016-12-23T19:55:18,22431,Cpanel-JSON-XS-3.0225| make: *** [XS.c] Error 255
```

**Comment by skaji** [comment]
Okay, I figured out why cpm sometimes failed to install Cpanel::JSON::XS.

cpm fails to install Cpanel::JSON::XS in the following step:

### 0.   
xsubpp is only in core `CORE/lib/perl5/5.10.1/ExtUtils/xsubpp`
ExtUtils::ParseXS in only in core `CORE/lib/perl5/5.10.1/ExtUtils/ParseXS.pm`
### 1.  
Execute `perl Makefile.PL` for Cpanel::JSON::XS,
so that Makefile are created, which uses xsubpp in core
### 2.  
Install a latest ExtUtils-ParseXS in local directory
### 3.  
Execute `make build` for Cpanel::JSON::XS,
then xsubpp in core is used, but it loads the latest ExtUtils::ParseXS from local
### 4.  
Ouch!

**Comment by skaji** [comment]
hopefully fixed by #67
